### ISSUE TITLE
Add collateral management/trading delegation

### ISSUE THREAD

**Issue by alnoki** [econia-labs/econia#2]
Presently, depositing/withdrawing collateral or placing trades requires a signature from a user, but this prohibits various forms of extensibility. A proposed solution is a delegated access approach, where users administer an on-chain authentication token to be used by other protocols. The proposed implementation involves the following uses and structs:

# Definitions

```move

use AptosFramework::Coin::Coin;
use AptosFramework::Table::Table;
use AptosFramework::TypeInfo::TypeInfo;
use Econia::CritBit::CritBit;

/// Representation of a `<B, Q, E>`-style market
struct Market has store {
    base_coin_type: TypeInfo,
    quote_coin_type: TypeInfo,
    scale_exponent: TypeInfo
}

/// Tracks all the markets that a user has accounts on
struct User has key {
    market_accounts: vector<Market>
}

/// Authentication token administered by a user when delegating access
struct AuthToken has store {
    user: address,
    market: Market,
    // For each user, each token has a unique serial ID
    serial_id: u64
}

/// Tracks a user's collateral, orders, and delegated access
struct MarketAccount<phantom B, phantom Q, phantom E> has key {
    base_coins: Coin<B>,
    base_coins_available: u64,
    quote_coins: Coin<Q>,
    quote_coins_available: u64,
    asks: CritBit<u64>,
    bids: CritBit<u64,
    scale_factor: u64,
    /// Counts the number of authentication tokens administered
    n_auth_tokens: u64,
    /// Map from authentication token to status: `true` for active,
    /// `false` for inactive
    auth_tokens: Table<u64, bool>
}

```

# Delegation 

Here, a user would delegate access via a function like:

```move
/// Return an `AuthToken`, which can be stored by third-party protocols
public(script) fun delegate_access<B, Q, E>(
    user: &signer
): AuthToken {
    // Get n_auth_tokens from user's market account, store as serial_id
    // Pack AuthToken with serial_id, <B, Q, E> Market, and user address
    // Add an entry to user's MarketAccount.auth_tokens
    // Increment MarketAccount.n_auth_tokens
    // Return the new AuthToken
}
```
# Verification

Here, third-party protocols could then call functions without a user's signature via functions of the form

```move
/// Perform a trade, withdraw or deposit collateral, etc,. for `user`,
/// provided `auth_token`
public fun function_requiring_token<B, Q, E>(
    user: address,
    auth_token: AuthToken
) {
    // Verify auth_token registered in user's MarketAccount.auth_tokens
    // Abort if user does not have access registered
    // Otherwise proceed
}
```

# Revoking access

Similar to giving someone else a password to a personal account, this strategy enables potentially-malicious behavior. As such, users should be easily able to revoke access for issued authentication tokens by simply changing the `bool` value in their `MarketAccount.auth_tokens` map to false:

```move
/// Revoke delegated access for `AuthToken` having `serial_id`
public(script) fun revoke_access<B, Q, E>(
    user: &signer
    serial_id: u64
) {
    // Borrow user's `MarketAccount`
    // Look up if `serial_id` in auth_tokens, flip value to `false`
}
```

**Comment by alnoki** [comment]
Regarding the proposal to grant the `store` ability to a `MarketAccount`, this strategy introduces an additional problem: if a third-party protocol stores a `MarketAccount` in an arbitrary struct, then the order book does no know where to route funds to in the case of a match.

And more generally, delegating permission introduces the additional issue of version number enforcement: if multiple delegated trades are placed from multiple users on the same market within the same transaction, users could potentially end up placing orders with the same version ID

**Comment by alnoki** [comment]
Additionally, a new `MarketAccount` should be created for each application with delegated permissions: if two protocols are granted withdraw permissions for a user's `MarketAccount`, and a liquidation event is about to occur, the two third-party protocols may come into contention over who gets to withdraw first.

**Comment by alnoki** [comment]
One issue with the implementation proposed directly above is that a signer could circumvent the counter by registering a `MarketAccount` without delegated custodianship, and a `MarketAccount` with custodianship delegated to themself. Hence they could place an order "on their own behalf" as a pseudo-custodian, and a conventional order with no delegated custodian, within the same transaction, and thus evade prior checks. Even if a user had a sequence number enforcement, this too could be circumvented by opening a separate account.

It would be possible to prevent this scenario by establishing a global counter for all orders within a single market, but this would introduce transaction collisions when submitting orders to paraqueues. Instead it is advised that order ID collisions be handled through separate means, for instance during the matching engine phase.

**Comment by alnoki** [comment]
To provide additional indexing functionality and visibility on collateral, it may also be useful to track the amount of collateral a user has within each given market, and total collateral, in a centralized user registry:

```move

/// Net collateral in a given market
struct MarketCollateral has store {
    /// Base coins held as collateral in a given market
    base_coin: u64,
    /// Quote coins held as collateral in a given market
    quote_coin: u64
}

/// Tracks markets that a user trades on and net collateral
struct UserMarketRegistry has key {
    /// All markets that a user has `MarketAccounts` for
    /// Map from `Market` type info tuple to collateral
    markets: open_table::OpenTable<Market, MarketCollateral> 
    /// Total collateral for all available coin types across all `MarketAccounts`
    net_collateral: open_table::OpenTable<TypeInfo, u64>
}
```

Here, when a user deposits or withdraws collateral, the `UserMarketRegistry` is updated accordingly, allowing cross-margin risk engines to inspect user info.

**Comment by alnoki** [comment]
Alternatively, the following data structures could minimize the number of required generic type parameters required for delegated collateral management:

```move
/// Market info defined in registry module
struct MarketInfo has copy, drop, store {
    /// Generic `CoinType` of `aptos_framework::coin::Coin`
    base_coin_type: type_info::TypeInfo,
    /// Generic `CoinType` of `aptos_framework::coin::Coin`
    quote_coin_type: type_info::TypeInfo,
    /// Scale exponent type defined in this registry module
    scale_exponent: type_info::TypeInfo
}

/// Info for a user's market account
struct MarketAccountInfo has copy, drop, store {
    /// The market that a user is trading on
    market_info: MarketInfo,
    /// Serial ID of a registered custodian, set to 0 when no authorized
    /// custodian for given market
    custodian_id: u64
}

/// Collateral for a given coin type, across different markets
struct Collateral<CoinType> has key {
    /// Map from market account info to the coins held as collateral on
    /// the given market. Separated into different table entries to
    /// prevent transaction collisions across markets
    accounts: open_table::open_table<MarketAccountInfo, Coin<CoinType>>
}

/// Represents a market account, book-keeping functionality only
struct MarketAccount has store {
    /// Collateral available for withdraw
    base_coins_available: u64,
    /// Collateral available for withdraw
    quote_coins_available: u64,
    /// Map from order ID to size of order, in base parcels
    asks: CritBit<u64>,
    /// Map from order ID to size of order, in base parcels
    bids: CritBit<u64>,
}

/// All of a user's market accounts across all markets
struct MarketAccounts has key {
    accounts: open_table::open_table<MarketAccountInfo, MarketAccount>
}
```

Here, collateral can be managed using a single `CoinType` generic for each `Collateral` resource table entry, with the corresponding `MarketAccountInfo` updated in `MarketAccounts` whenever a deposit, withdraw, or order is made. This architecture does not break parallelism, and allows delegated custodians to manage collateral by passing only one generic type parameter.

**Comment by alnoki** [comment]
Closed per v2.0.0
